tests/tests/compiler_docs.rs

```diff
@@ -0,0 +1,169 @@
+            vec
+        } else {
+            self.collected.insert(key.clone(), Vec::new());
+            self.collected.get_mut(&key).unwrap()
+        };
+
+        vec.push(doc.to_string());
+    }
+
+}
+
+impl DocVisitor {
+
+    fn assert(&self) {
+        for (k, expected) in &self.expected {
+            let against = if let Some(vec) = self.collected.get(*k) {
+                vec
+            } else {
+                panic!("missing documentation for item {k:?}");
+            };
+
+            let mut idx: usize = 0;
+            while idx < expected.len() {
+                if let Some(collected) = against.get(idx) {
+                    assert_eq!(collected, expected[idx], "mismatched docstring:\n\tgot: {:?}\n\texp: {:?}", collected, expected[idx]);
+                } else {
+                    panic!("missing docstrings, expected: {:?}", expected[idx]);
+                }
+
+                idx += 1;
+            }
+
+            if expected.len() < against.len() {
+                let (_, extras) = against.split_at(expected.len());
+                panic!("extra docstrings: {:?}", extras);
+            }
+        }
+
+        if self.collected.len() > self.expected.len() {
+            let vec = self.collected.keys().filter(|it| !self.expected.contains_key(it.as_str())).collect::<Vec<_>>();
+            panic!("encountered more documented items than expected: {vec:?}");
+        }
+    }
+
+}
+
+macro_rules! expect_docs {
+    ($($typename:literal => { $($docstr:literal)* })+) => {
+        {
+            let mut expected = BTreeMap::new();
+
+            $(
+            #[allow(unused_mut)]
+            let mut vec = Vec::new();
+            $(vec.push($docstr);
+            )*
+
+            expected.insert($typename, vec);
+            )+
+
+            DocVisitor {
+                expected,
+                collected: BTreeMap::new()
+            }
+        }
+    };
+}
+
+#[test]
+fn harvest_docs() {
+    let mut diagnostics = Diagnostics::new();
+    let mut vis = expect_docs! {
+        "stuff" => { " Top-level function.\n" }
+        "Struct" => {
+            " Top-level struct.\n"
+            " Second line!\n"
+        }
+        "Enum" => { "\n         * Top-level enum.\n         " }
+        "Enum::A" => { " Enum variant A.\n" }
+        "Enum::B" => { " Enum variant B.\n" }
+        "constant" => { " Top-level constant.\n" }
+
+        // Curiously, the module item itself never get registered. This always fails.
+        // The child items of the module are, though.
+        // "Module" => { " Top-level module.\n" }
+
+        "Module::Enum" => { " Module enum.\n" }
+        "Module::Enum::A" => { " Enum variant A.\n" }
+        "Module::Enum::B" => { " Enum variant B.\n" }
+    };
+
+    let mut sources = sources(r#"
+        //! Mod/crate doc.
+        /*! Multiline mod/crate doc.
+         *
+         *  Both of these comments disappear; we can't hold onto file-level attributes at the moment..
```

Comment:
[New issue?](https://github.com/rune-rs/rune/blob/0ed2e7db04b20206c3cb18f0ccac2a63da16c287/crates/rune/src/indexing/index.rs#L585)

Reply:
Yeah, this PR is in a decent state so lets break it into a separate one.

Capturing file and module level documentation requires adding calls to the associated visitor.